### PR TITLE
service: Filter several services using config

### PR DESCRIPTION
**Pull request by gvso**
This uses the targets configuration and concurrently filters the services based on the selectors.

### CODE REVIEW

File: internal/run/wrapper.go [hunk 1]
```diff
@@ -21,6 +21,22 @@ type API struct {
 	Region string
 }
 
+// Regions are the available regions.
+var Regions = []string{
```
Comment (ahmetb) [inline internal/run/wrapper.go:25]: this can be fetched from api.

File: cmd/operator/main.go [hunk 2]
```diff
@@ -119,19 +117,10 @@ func main() {
...
-		Region:      flRegions[0],
-	}, cfg.Strategy).WithLogger(logger)
-
 	for {
+		services := service.Filter(cfg)
+		roll := rollout.New(services[0], cfg.Strategy).WithLogger(logger)
+
 		changed, err := roll.Rollout()
 		if err != nil {
 			logger.Infof("Rollout failed: %v", err)
```
Comment (ahmetb) [inline cmd/operator/main.go:126]: re: this code block (you can address in a different PR)

```
		interval := time.Duration(cfg.Strategy.Interval)
		time.Sleep(interval * time.Second)

```

we need to find a better name for "Interval". (e.g. "What interval is this for?")

File: internal/run/wrapper.go [hunk 3]
```diff
@@ -47,6 +63,12 @@ func (a *API) ReplaceService(namespace, serviceID string, svc *run.Service) (*ru
 	return a.Client.Namespaces.Services.ReplaceService(serviceName, svc).Do()
 }
 
+// ListServices gets services filtered by a label.
+func (a *API) ListServices(namespace string, labelSelector string) (*run.ListServicesResponse, error) {
+	parent := fmt.Sprintf("namespaces/%s", namespace)
+	return a.Client.Namespaces.Services.List(parent).LabelSelector(labelSelector).Do()
```
Comment (ahmetb) [inline internal/run/wrapper.go:69]: interesting that we support server-side label selector!

File: cmd/operator/main.go (hunk 2 above)
Comment (gvso) [reply cmd/operator/main.go:131]: Yes, we should probably make it clearer. We can change it in another PR since this interval is only used here.

Comment (ahmetb) [inline cmd/operator/main.go:131]: This probably should be fatalf.

File: cmd/operator/main.go [hunk 4]
```diff
@@ -119,19 +117,15 @@ func main() {
...
-		RunClient:   runclient,
-		Project:     cfg.Targets[0].Project,
-		ServiceName: "hello",
-		Region:      flRegions[0],
-	}, cfg.Strategy).WithLogger(logger)
-
 	for {
+		services := service.Filter(logger, cfg)
+		if len(services) == 0 {
+			logger.Info("No service matches the targets")
```
Comment (ahmetb) [inline cmd/operator/main.go:123]: this is probably a warn

File: cmd/operator/main.go [hunk 5]
```diff
@@ -119,19 +117,15 @@ func main() {
...
+			interval := time.Duration(cfg.Strategy.Interval)
```
Comment (ahmetb) [inline cmd/operator/main.go:124]: try not to repeat code. this part is repeated right now

**Comment by ahmetb** [comment]
cc: @grayside if you have time to take a look. this PR is about fetching and filtering Cloud Run services from the API.

File: pkg/service/service.go [hunk 6]
```diff
@@ -1,13 +1,125 @@
...
+		RunClient:   runclient,
+		Service:     svc,
+		Project:     project,
+		ServiceName: serviceName,
+		Region:      region,
+	}
+}
+
+// Filter retrieves all the services that match the filter in the configuration.
+func Filter(logger *logrus.Logger, cfg *config.Config) []*Client {
```
Comment (ahmetb) [inline pkg/service/service.go:35]: 1. I think this method is poorly named. It sounds like QueryAndFilterAllServices()

2. You only use `cfg.Targets` so why are you passing `config.Config` here?

File: pkg/service/service.go [hunk 7]
```diff
@@ -1,13 +1,125 @@
...
+	clientsCh := make(chan []*Client)
```
Comment (ahmetb) [inline pkg/service/service.go:36]: this code is hard to follow wit channels, I'm not sure they are required here. what are you trying to do? 

if all you're trying to do is get services in parallel, you should use WaitGroup + Goroutines; not channel.

File: pkg/service/service.go [hunk 8]
```diff
@@ -1,13 +1,125 @@
...
+	numberOfFilterByRegionCalls := 0
```
Comment (ahmetb) [inline pkg/service/service.go:37]: `filterByRegionCall` identifier also doesn't make to me, isn't the task here simply:

1. Identify regions (not this method)
2. Choose which regions to use if a region list is specified (not this method)
3. Query all services in all regions "in parallel" + wait
4. As part of (3), query services in 1 region, with label filter.

so I'm not sure which one of 1,2,3,4 I'm reading right now. It seems it's sort of all. :)

File: pkg/service/service.go [hunk 9]
```diff
@@ -1,13 +1,125 @@
...
+		"project":        project,
+		"region":         region,
+		"selectorType":   selector.Type,
+		"selectorFilter": selector.Filter,
+	})
+
+	runclient, err := runapi.NewAPIClient(context.Background(), region)
+	if err != nil {
+		lg.Errorf("can not initialize Cloud Run client: %v", err)
+		clientsCh <- nil
```
Comment (ahmetb) [inline pkg/service/service.go:83]: don't use channels, they aren't required.

File: pkg/service/service.go [hunk 10]
```diff
@@ -1,13 +1,125 @@
...
+	}
+
+	close(clientsCh)
+
+	return svcClients
+}
+
+// filterByRegion searches for a service name or label selector in an specific
+// regional endpoint.
+func filterByRegion(clientsCh chan []*Client, logger *logrus.Logger, project, region string, selector config.TargetSelector) {
```
Comment (ahmetb) [inline pkg/service/service.go:72]: again this name isn't good. I can't understand what this method by looking at the implementation.

I see "service name type" but we said we'll do label based selectors. I'm not entirely sure what's happening here. Furthermore `chan`s just contribute to the complexity.

**Comment by ahmetb** [comment]
Right now, I'm seeing overcomplicated code in this PR getting close to spaghetti. :) It would be good to go to the drawing board and rethink some of these packages and split them into smaller methods.

Now that the algorithms are getting hard to follow, it usually points to _code smell_. What we are trying to do here is fairly simple: Query all services in some/all regions with a label selector. That's it. But the code is far beyond that.

File: pkg/service/service.go [hunk 11]
```diff
@@ -1,13 +1,125 @@
...
-// Client is the connection to update and obtain data about a single service.
+// Client is the connection to update and obtain data about a single service. It
+// contains a client to the Cloud Run API, information about the service and the
+// service object.
 type Client struct {
 	RunClient   runapi.Client
+	Service     *run.Service
 	Project     string
 	ServiceName string
 	Region      string
```
Comment (grayside) [inline pkg/service/service.go:20]: Given project, name, and region are all things that can be found on the service object, is there a value in duplicate data? If not, maybe these would be better served by helper methods.

File: pkg/service/service.go (hunk 7 above)
Comment (gvso) [reply pkg/service/service.go:36]: The major reason this code looks confusing is that I'm using channels to retrieve the service objects I am getting from the API. If I run this in goroutines, what's the best way to store the returned values from the API using a WaitGroup? Using global state? Creating a struct with a `[]run.Service/[]service.Client` field + sync.WaitGroup field? Passing the array storing the values and using mutex?

**Comment by gvso** [comment]
@ahmetb The major reason this code looks confusing is that I'm using channels to retrieve the service objects I am getting from the API. If I run this in goroutines, what's the best way to store the returned values from the API using a WaitGroup? Using global state? Creating a struct with a []run.Service/[]service.Client field + sync.WaitGroup field? Passing the array storing the values and using mutex?

The sources I have seen either don't have my use case/use the channel implementation. I'd appreciate any guidelines here since I haven't done concurrency in real-world applications (only at school labs).

File: cmd/operator/main.go [hunk 12]
```diff
@@ -119,19 +117,15 @@ func main() {
...
+			time.Sleep(interval * time.Second)
+		}
+		roll := rollout.New(services[0], cfg.Strategy).WithLogger(logger)
```
[the remark this answers is not in the document]
Comment (gvso) [reply cmd/operator/main.go:127]: This is definitively a TODO. I didn't want to overcomplicate this PR since rollout of several services would be in another PR. I'll add a TODO here. Thanks!

File: pkg/service/service.go [hunk 13]
```diff
@@ -1,13 +1,125 @@
...
+	"context"
+
 	runapi "github.com/GoogleCloudPlatform/cloud-run-release-operator/internal/run"
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/pkg/config"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/run/v1"
 )
 
-// Client is the connection to update and obtain data about a single service.
+// Client is the connection to update and obtain data about a single service. It
```
[the remark this answers is not in the document]
Comment (gvso) [reply pkg/service/service.go:12]: That's possible. However, just take into consideration that we probably want to run the rollout process concurrently, so we need a `service.Client` per filtered service.

That being said, I would probably need to create a cache of `run.Client`s for regions the filter has seen already. Is it worth doing so in this PR or should we do it in another one?

File: pkg/service/service.go (hunk 11 above)
Comment (gvso) [reply pkg/service/service.go:20]: I'm not sure that all these data can be extracted from the service object. Looking at the YAML in the Cloud Run service, I get

```yaml
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: hello
  namespace: '929898821634'
  generation: 130
  creationTimestamp: '2020-06-29T22:14:02.259763Z'
  labels:
    rollout-strategy: gradual
    cloud.googleapis.com/location: us-east1
  annotations:
    run.googleapis.com/client-name: gcloud
    rollout.cloud.run/stableRevision: hello-00124-cim
    client.knative.dev/user-image: gcr.io/cloudrun/hello
    run.googleapis.com/client-version: 298.0.0
    run.googleapis.com/launch-stage: ALPHA
```
Can we trust that all types of deployment (not using gcloud, for instance) are going to include the `name` and `namespace` metadata? The region looks like a label that was created by my deployment tool (gcloud).

**Comment by ahmetb** [comment]
The only variable in these parallel executions is region name. So you don't need an array of client instances, you can init them in Goroutines. Would be good to study some sync.WorkGroup examples on the internet. You can simply append the resulting svc records into an array.

File: pkg/service/service.go (hunk 11 above)
Comment (ahmetb) [reply pkg/service/service.go:20]: name/namespace are mandatory.
however don't rely on Google Cloud specific annotations since we might support knative someday.

If you keep passing `<region,name,project>` triplet around, just create a `ServiceRecord` struct?

File: pkg/service/service.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (gvso) [reply pkg/service/service.go:12]: After discussing it with Ahmet, I think there's no harm in creating as many `run.Client`s as needed since initializing them does not make any HTTP requests per sé. So, there's no connection that should be closed or anything like that, and it's not using much memory. Therefore, I think it's better to creating the object as needed to increase code readability.

File: cmd/operator/service.go [hunk 14]
```diff
@@ -0,0 +1,92 @@
+package main
```
Comment (gvso) [inline cmd/operator/service.go:1]: Should I keep this file or just move the content to the `main.go` file?

Comment (ahmetb) [reply cmd/operator/service.go:1]: We can move. Doesn't make a lot of difference for now.

File: cmd/operator/service.go [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+	}
+
+	var records []*rollout.ServiceRecord
+	for _, svc := range svcs {
+		records = append(records, &rollout.ServiceRecord{
+			Service: svc,
+			Project: project,
+			Region:  region,
+		})
+	}
```
Comment (ahmetb) [inline cmd/operator/service.go:78]: you prob need a ToServiceRecord method to call from goroutine, this doesn't have to be here.

File: cmd/operator/service.go [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+
+	return records, nil
+}
+
+func getRegions(target *config.Target) ([]string, error) {
```
Comment (ahmetb) [inline cmd/operator/service.go:83]: add comment

File: cmd/operator/service.go [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+	regions := target.Regions
+	if len(regions) == 0 {
```
Comment (ahmetb) [inline cmd/operator/service.go:85]: flip this !=

File: internal/run/wrapper.go [hunk 18]
```diff
@@ -47,6 +50,44 @@ func (a *API) ReplaceService(namespace, serviceID string, svc *run.Service) (*ru
...
+	var services []*run.Service
+	for _, service := range servicesList.Items {
+		services = append(services, service)
+	}
+	return services, nil
+}
+
+// Regions gets the supported regions for the project.
+func Regions(project string) ([]string, error) {
+	if len(regions) == 0 {
```
Comment (ahmetb) [inline internal/run/wrapper.go:71]: flip !=

Comment (ahmetb) [reply internal/run/wrapper.go:71]: add logs about using cached regions

File: internal/util/types.go [hunk 19]
```diff
@@ -0,0 +1,8 @@
+package util
+
+type contextKey string
+
+// Context keys.
+const (
```
Comment (gvso) [inline internal/util/types.go:6]: This is used in only one place as of now, but it'll be useful when adding the health package and logs in metrics/stackdriver

File: cmd/operator/service.go [hunk 20]
```diff
@@ -83,18 +88,22 @@ func getServicesByRegionAndLabel(ctx context.Context, logger *logrus.Logger, pro
 // If the target configuration does not specify any regions, the entire list of
 // regions is retrieved from API.
 func determineRegions(ctx context.Context, logger *logrus.Logger, target *config.Target) ([]string, error) {
+	lg := logger.WithField("labelSelector", target.LabelSelector)
```
Comment (ahmetb) [inline cmd/operator/service.go:91]: label selector is not applicable to determining regions. so probably this log stmt shouldn't have this field

File: cmd/operator/main.go [hunk 21]
```diff
@@ -119,32 +110,40 @@ func main() {
...
 			logger.Info("Rollout process succeeded")
 		}
 
-		interval := time.Duration(cfg.Strategy.Interval)
-		time.Sleep(interval * time.Second)
+		sleepInSeconds(cfg.Strategy.Interval)
 	}
 }
 
+func sleepInSeconds(interval int64) {
```
Comment (ahmetb) [inline cmd/operator/main.go:142]: no need to create this method. this is already 1-liner

File: cmd/operator/service.go [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+		retServices []*rollout.ServiceRecord
+		retError    error
+		mu          sync.Mutex
+		wg          sync.WaitGroup
+	)
+
+	for _, target := range targets {
+		regions, err := determineRegions(ctx, logger, target)
+		if err != nil {
+			logger.Errorf("cannot determine regions: %v", err)
```
Comment (ahmetb) [inline cmd/operator/service.go:33]: this is probably a `return errors.Wrap`.

fail early fail often.

File: cmd/operator/service.go [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+
+// getServicesByRegionAndLabel returns all the service records that match the
+// labelSelector in a specific region.
+func getServicesByRegionAndLabel(ctx context.Context, logger *logrus.Logger, project, region, labelSelector string) ([]*run.Service, error) {
+	lg := logger.WithFields(logrus.Fields{
+		"region":        region,
+		"labelSelector": labelSelector,
+	})
+
+	lg.Debugf("filtering services")
```
Comment (ahmetb) [inline cmd/operator/service.go:71]: this is incorrect because you don't have any services to filter yet. :)

Comment (ahmetb) [reply cmd/operator/service.go:71]: `querying Cloud Run services` would be an appropriate statement

File: cmd/operator/service.go [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+
+// determineRegions gets the regions the label selector should be searched at.
+//
+// If the target configuration does not specify any regions, the entire list of
+// regions is retrieved from API.
+func determineRegions(ctx context.Context, logger *logrus.Logger, target *config.Target) ([]string, error) {
+	lg := logger.WithField("labelSelector", target.LabelSelector)
+	regions := target.Regions
+	if len(regions) != 0 {
+		lg.Debug("regions set in configuration")
```
Comment (ahmetb) [inline cmd/operator/service.go:94]: `using predefined list of regions, skip querying from API` is a more descriptive

File: internal/util/types.go [hunk 25]
```diff
@@ -0,0 +1,8 @@
...
+	ContextKeyLogger = contextKey("logger")
```
Comment (ahmetb) [inline internal/util/types.go:7]: this pkg is nor useful right now.

skip for now. create new PR where you properly design a `util.WithLogger(ctx, logger)` method.

ContextKeyLogger is an implementation detail that should not be exposed to the outside world.

File: cmd/operator/main.go [hunk 26]
```diff
@@ -119,32 +110,40 @@ func main() {
 }
 
 func runCLI(logger *logrus.Logger, cfg *config.Config) {
```
Comment (ahmetb) [inline cmd/operator/main.go:112]: this probably can return an `error`. use `log.fatal`s sparingly because they ignore all call stack and finish the process immediately.

if you use `return`s as much as we can, we can do `log.fatal` in max 1-2 places, and have a chance to fail with either just the error message and potentially also with the callstack in debug mode.

File: internal/run/wrapper.go [hunk 27]
```diff
@@ -21,6 +21,9 @@ type API struct {
 	Region string
 }
 
+// regions are the available regions.
+var regions = []string{}
```
Comment (ahmetb) [inline internal/run/wrapper.go:25]: can you add a `// TODO caching regions might be unnecessary if we are querying them once during the lifespan of the process`

File: internal/run/wrapper.go [hunk 28]
```diff
@@ -47,6 +50,41 @@ func (a *API) ReplaceService(namespace, serviceID string, svc *run.Service) (*ru
...
+// ServicesWithLabelSelector gets services filtered by a label selector.
+func (a *API) ServicesWithLabelSelector(namespace string, labelSelector string) ([]*run.Service, error) {
+	parent := fmt.Sprintf("namespaces/%s", namespace)
+
+	servicesList, err := a.Client.Namespaces.Services.List(parent).LabelSelector(labelSelector).Do()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to filter services by label selector")
+	}
+
+	return servicesList.Items, nil
```
Comment (ahmetb) [inline internal/run/wrapper.go:62]: this is something to learn, you can do:

`return servicesList.Items, errors.Wrap(err, "..")`

Wrap returns nil if err is nil. But make sure `servicesList` isn't a pointer type that can panic in case of error.

File: internal/run/wrapper.go [hunk 29]
```diff
@@ -47,6 +50,41 @@ func (a *API) ReplaceService(namespace, serviceID string, svc *run.Service) (*ru
...
+}
+
+// Regions gets the supported regions for the project.
+func Regions(project string) ([]string, error) {
+	if len(regions) != 0 {
+		return regions, nil
+	}
+
+	client, err := run.NewService(context.Background())
```
Comment (ahmetb) [inline internal/run/wrapper.go:71]: avoid `context.Background` pass `ctx` here. That's why we have ctx, if we don't use it, it's pointless.

File: internal/run/wrapper.go (hunk 28 above)
Comment (gvso) [reply internal/run/wrapper.go:62]: That's helpful and I didn't know about it. I wonder if in this case it is better as it is since `servicesList` is a pointer, so I will have add an if anyways. Since is more common to see check for errors, I think an `if err != nil` make it more readable.

Comment (ahmetb) [reply internal/run/wrapper.go:62]: If servicesList.Items has a chance of panicking, don't do what I said. :) it works otherwise however

File: cmd/operator/main.go (hunk 26 above)
Comment (gvso) [reply cmd/operator/main.go:112]: That's a good point. I'll change it once we start handling several services. The `log.Fatal` when rollout fails will go away since rollout of service A shouldn't probably affect rollout of service B. 

But yeah, there are more cases where we can error entirely (e.g. fail to filter targeted services)

File: cmd/operator/service.go [hunk 30]
```diff
@@ -76,13 +76,13 @@ func getServicesByRegionAndLabel(ctx context.Context, project, region, labelSele
 //
 // If the target configuration does not specify any regions, the entire list of
```
Comment (ahmetb) [inline cmd/operator/service.go:77]: fix in a separate PR please:
this is not proper godoc. godocs start with the method's name.
